README.md
# 分支：[淘客助手](https://github.com/ouqinglai/12306buy/tree/taobaobuy) [12306订票助手](https://github.com/ouqinglai/12306buy) [12306旧版(弃)](https://github.com/ouqinglai/12306buy/tree/old_dama)

> 12306buy订票助手：chrome extension plugin 浏览器插件
>
> 一键下单，自动通过验证码，带循环查票功能。后台运行，桌面通知
>
> **browser :** 请使用最新360极速浏览器.
>
> **passenger :** 请先到[12306](https://kyfw.12306.cn/otn/passengers/addInit)设置一个常用联系人. 
>
> **声明 :** 感谢360团队破解12306验证码(helper.dll)，仅作学习用途. 

## step

### login 12306 (登录太慢时，可到12306登录页面手动登录)
![login 12306](https://github.com/ouqinglai/12306buy/blob/master/screenshot/login.png)

### write ticket information
![main screen](https://github.com/ouqinglai/12306buy/blob/master/screenshot/main.png)

## chrome.notifications(桌面通知)

### when has ticket
![when has ticket](https://github.com/ouqinglai/12306buy/blob/master/screenshot/hasTicket.png)

### tip box
![tip box](https://github.com/ouqinglai/12306buy/blob/master/screenshot/tip.png)

### success!!
![addOrder success](https://github.com/ouqinglai/12306buy/blob/master/screenshot/success.png)


CHANGELOG.md
## 1.42

`2016-12-29`

* 优化查票接口，改用去哪儿api

## 1.41

`2016-12-28`

* 修改询票时，只关心能预定.
* 优化是否已登录查询方法.
* 新增fetch API timeout功能.

## 1.40

`2016-12-27`

* 修改查询接口名称，并改为自动.
* 优化站名设为本地文件.

## 1.30

`2016-12-23`

* 新增出发/目的地站名选择列表.
* 修复loading、msg === ""循环时间

## 1.20

`2016-12-16`

* 提交订单不再需要验证码
* 更新360的helper.dll


js/popup.js
var url1 = 'https://kyfw.12306.cn/otn/'
,	user = JSON.parse(localStorage.USER || '{ "user" : "" , "pwd" : "" }')//user和pwd字段固定
,	stationName
,	isSelectStation = {
	_from : false,//出发地或目的地是否已选中值，false时不能提交form
	_to : false,
}
,	hook = [e => e.preventDefault() , (match , value) => {
	chrome.runtime.sendMessage({
        match,
        value,
    })
}]
,	$from = $('input[name="from"]')
,	$to = $('input[name="to"]')
,	$submit = $('#order_submit')

/* init */
$('[name="to_date"]').value = new Date().toISOString().split('T')[0]
$('[name="time"]').value = new Date().toTimeString().match(/\d+:\d+/)[0]

//读取12306账户信息
$('[name="user"]').value = user.user
$('[name="pwd"]').value = user.pwd

//check 12306 login
Fetch('login/checkUser' , res => whichFormShow(res.data.flag ? 3 : 2))

//当错误提示框后台运行时，点击图标时让它再次显示
chrome.runtime.sendMessage({ match : 'iconClick' })

//获取车站对应编码
fetch(chrome.extension.getURL('js/stationName.txt'))
.then(res => res.text())
.then(res => {
	stationName = res

	isSelectStation = {
		_from : !!res.match($from.value),
		_to : !!res.match($to.value),
	}
})

/* eventBind */
$('#loginForm').onsubmit = e => {
	hook[0](e)

	hook[1]('userLogin' , formData(this))
}

$to.onblur = $from.onblur = function (){
	let $select = this.nextElementSibling

	if((new RegExp(`<li>${ this.value }<\/li>`)).test($select.outerHTML)) {
		isSelectStation['_' + this.getAttribute('name')] = true
	}
	setTimeout(() => {
		$select.style.display = 'none'
	} , 100)
}

$to.onkeyup = $to.onclick = $from.onkeyup = $from.onclick = function (e) {
	let _inputSelf = this
	,	val = _inputSelf.value
	,	inputName = _inputSelf.getAttribute('name')
	,	nextSelectDom = _inputSelf.nextElementSibling
	,	wrapperDom = document.createDocumentFragment()

	e.type !== 'click' && (isSelectStation['_' + inputName] = false)

	if(val) {
		nextSelectDom.innerHTML = '<span class="no-more">查找无效~</span>'

		let match = stationName.match(new RegExp(`\\|${ val }\\W*\\|` , 'g'))
		match && match.forEach(name => {
			let li = document.createElement('li')

			li.innerHTML = name.replace(/\|/g , '')
			li.onclick = function (){
				nextSelectDom.style.display = 'none'

				isSelectStation['_' + inputName] = true

				_inputSelf.value = this.innerHTML
			}

			wrapperDom.appendChild(li)
		})

		nextSelectDom.insertBefore(wrapperDom , nextSelectDom.childNodes[0])
		nextSelectDom.style.display = 'block'
	}
}

$('#order').onsubmit = e => {
	hook[0](e)

	if(!isSelectStation._from) return alert('出发地无效！')
	if(!isSelectStation._to) return alert('目的地无效！')

	let orderInfo = formData(this)
	,	{ from , to } = orderInfo

	orderInfo.from = findStationCode(from)
	orderInfo.to = findStationCode(to)
	orderInfo.user = user

	if(orderInfo.mode === 'buy') $submit.setAttribute('disabled' , '')

	hook[1]('orderInfo' , orderInfo)
}

chrome.runtime.onMessage.addListener(({ match , value }) => {
	if(match === 'loginCb') {
		localStorage.USER = JSON.stringify(value)
		whichFormShow(3)
	}else if (match === 'errorCb') {
		$submit.removeAttribute('disabled')
	}else if (match === 'loading') {
		$$forEach('button[type="submit"]' , dom => dom.classList.toggle('loading'))
	}
})

/* helper */

//获取常用联系人（乘客）信息
function getPassenger (){
	Fetch('confirmPassenger/getPassengerDTOs' , ({ data : { normal_passengers } }) => {
		let passenger = normal_passengers[0]

		if(passenger) {
			Array('passenger_name' , 'passenger_id_no' , 'mobile_no').forEach(id => $('#' + id).value = passenger[id] || '(无)')
		}else {
			$submit.setAttribute('disabled' , '')
			new Notification('请先设置常用联系人！' , {
				icon : './logo.png'
			})
		}
	})
}

function Fetch (api , cb , type = 'json') {
	let promise = fetch(url1 + api , {
		credentials: 'include',
		headers : { 'Cache-Control' : 'no-cache' }
	})

	if(cb) {
		promise
		.then(res => res[type]())
		.then(res => cb(res))
	}else
		return promise
}

function findStationCode (name){
	return [name , stationName.match(new RegExp('\\|' + name + '\\|[a-zA-z]+\\|'))[0].split('|')[2]]
}

function $ (selector){
	return document.querySelector(selector)
}

function $$forEach (selector , cb){
	[].forEach.call(document.querySelectorAll(selector) , cb)
}

function formData (form){
	let obj = {}

	$$forEach('[name]' , dom => {
		if(dom.type === 'radio' || dom.type === 'checkbox') {
			dom.checked && (obj[dom.name] = dom.value)
		}else if(dom.name) obj[dom.name] = dom.value
	})

	return obj
}

function whichFormShow (index) {
	if(index === 3) getPassenger()

	document.body.setAttribute('class' , 'form' + index)
}

js/update.js
/*一天检测一次*/
var _date = (new Date).toLocaleDateString()

localStorage.getItem('updata_version') !== _date && chrome.management.getSelf(({ updateUrl , version }) => {
	fetch(updateUrl)
	.then(res => res.text())
	.then(res => {
		localStorage.setItem('updata_version' , _date)

		let new_version = res.match(new RegExp('</span>version<span.+</span>,</td>'))[0].match(/([\.\d]+)/)[0]

		if(version < new_version) {
			fetch(updateUrl.replace('manifest.json' , 'CHANGELOG.md'))
			.then(text => text.text())
			.then(text => {
				let changeLog = []

				text
				.match(new RegExp(`<\/a>${ new_version }<\/h2>(\\s\|\\S)+<\/ul>\\s<\/article>`))[0]
				.match(/<ul>\s(<li>.+<\/li>\s)+<\/ul>/)[0]
				.match(/<li>.+<\/li>/g)
				.forEach((li , index) => {
					changeLog.push({
						title : index + 1 + '.',
						message : li.replace(/(<li>|<\/li>)/g , '')
					})
				})
				
				chrome.notifications.create('' , {
					type: 'list',
			        title: `新版本${ new_version }升级提示！`,
			        message: `msg`,
			        iconUrl: './logo.png',
			        items: changeLog,
			        requireInteraction : true,//一直显示
				})
			})
		}
	})
})

js/taobaosign.js
/*解密淘宝sign函数*/

function taobaoSign(a) {
   function b(a, b) {
       return a << b | a >>> 32 - b
   }

   function c(a, b) {
       var c, d, e, f, g;
       return e = 2147483648 & a, f = 2147483648 & b, c = 1073741824 & a, d = 1073741824 & b, g = (1073741823 & a) + (1073741823 & b), c & d ? 2147483648 ^ g ^ e ^ f : c | d ? 1073741824 & g ? 3221225472 ^ g ^ e ^ f : 1073741824 ^ g ^ e ^ f : g ^ e ^ f
   }

   function d(a, b, c) {
       return a & b | ~a & c
   }

   function e(a, b, c) {
       return a & c | b & ~c
   }

   function f(a, b, c) {
       return a ^ b ^ c
   }

   function g(a, b, c) {
       return b ^ (a | ~c)
   }

   function h(a, e, f, g, h, i, j) {
       return a = c(a, c(c(d(e, f, g), h), j)), c(b(a, i), e)
   }

   function i(a, d, f, g, h, i, j) {
       return a = c(a, c(c(e(d, f, g), h), j)), c(b(a, i), d)
   }

   function j(a, d, e, g, h, i, j) {
       return a = c(a, c(c(f(d, e, g), h), j)), c(b(a, i), d)
   }

   function k(a, d, e, f, h, i, j) {
       return a = c(a, c(c(g(d, e, f), h), j)), c(b(a, i), d)
   }

   function l(a) {
       for (var b, c = a.length, d = c + 8, e = (d - d % 64) / 64, f = 16 * (e + 1), g = new Array(f - 1), h = 0, i = 0; c > i;) b = (i - i % 4) / 4, h = i % 4 * 8, g[b] = g[b] | a.charCodeAt(i) << h, i++;
       return b = (i - i % 4) / 4, h = i % 4 * 8, g[b] = g[b] | 128 << h, g[f - 2] = c << 3, g[f - 1] = c >>> 29, g
   }

   function m(a) {
       var b, c, d = "",
           e = "";
       for (c = 0; 3 >= c; c++) b = a >>> 8 * c & 255, e = "0" + b.toString(16), d += e.substr(e.length - 2, 2);
       return d
   }

   function n(a) {
       a = a.replace(/\r\n/g, "\n");
       for (var b = "", c = 0; c < a.length; c++) {
           var d = a.charCodeAt(c);
           128 > d ? b += String.fromCharCode(d) : d > 127 && 2048 > d ? (b += String.fromCharCode(d >> 6 | 192), b += String.fromCharCode(63 & d | 128)) : (b += String.fromCharCode(d >> 12 | 224), b += String.fromCharCode(d >> 6 & 63 | 128), b += String.fromCharCode(63 & d | 128))
       }
       return b
   }
   var o, p, q, r, s, t, u, v, w, x = [],
       y = 7,
       z = 12,
       A = 17,
       B = 22,
       C = 5,
       D = 9,
       E = 14,
       F = 20,
       G = 4,
       H = 11,
       I = 16,
       J = 23,
       K = 6,
       L = 10,
       M = 15,
       N = 21;
   for (a = n(a), x = l(a), t = 1732584193, u = 4023233417, v = 2562383102, w = 271733878, o = 0; o < x.length; o += 16) p = t, q = u, r = v, s = w, t = h(t, u, v, w, x[o + 0], y, 3614090360), w = h(w, t, u, v, x[o + 1], z, 3905402710), v = h(v, w, t, u, x[o + 2], A, 606105819), u = h(u, v, w, t, x[o + 3], B, 3250441966), t = h(t, u, v, w, x[o + 4], y, 4118548399), w = h(w, t, u, v, x[o + 5], z, 1200080426), v = h(v, w, t, u, x[o + 6], A, 2821735955), u = h(u, v, w, t, x[o + 7], B, 4249261313), t = h(t, u, v, w, x[o + 8], y, 1770035416), w = h(w, t, u, v, x[o + 9], z, 2336552879), v = h(v, w, t, u, x[o + 10], A, 4294925233), u = h(u, v, w, t, x[o + 11], B, 2304563134), t = h(t, u, v, w, x[o + 12], y, 1804603682), w = h(w, t, u, v, x[o + 13], z, 4254626195), v = h(v, w, t, u, x[o + 14], A, 2792965006), u = h(u, v, w, t, x[o + 15], B, 1236535329), t = i(t, u, v, w, x[o + 1], C, 4129170786), w = i(w, t, u, v, x[o + 6], D, 3225465664), v = i(v, w, t, u, x[o + 11], E, 643717713), u = i(u, v, w, t, x[o + 0], F, 3921069994), t = i(t, u, v, w, x[o + 5], C, 3593408605), w = i(w, t, u, v, x[o + 10], D, 38016083), v = i(v, w, t, u, x[o + 15], E, 3634488961), u = i(u, v, w, t, x[o + 4], F, 3889429448), t = i(t, u, v, w, x[o + 9], C, 568446438), w = i(w, t, u, v, x[o + 14], D, 3275163606), v = i(v, w, t, u, x[o + 3], E, 4107603335), u = i(u, v, w, t, x[o + 8], F, 1163531501), t = i(t, u, v, w, x[o + 13], C, 2850285829), w = i(w, t, u, v, x[o + 2], D, 4243563512), v = i(v, w, t, u, x[o + 7], E, 1735328473), u = i(u, v, w, t, x[o + 12], F, 2368359562), t = j(t, u, v, w, x[o + 5], G, 4294588738), w = j(w, t, u, v, x[o + 8], H, 2272392833), v = j(v, w, t, u, x[o + 11], I, 1839030562), u = j(u, v, w, t, x[o + 14], J, 4259657740), t = j(t, u, v, w, x[o + 1], G, 2763975236), w = j(w, t, u, v, x[o + 4], H, 1272893353), v = j(v, w, t, u, x[o + 7], I, 4139469664), u = j(u, v, w, t, x[o + 10], J, 3200236656), t = j(t, u, v, w, x[o + 13], G, 681279174), w = j(w, t, u, v, x[o + 0], H, 3936430074), v = j(v, w, t, u, x[o + 3], I, 3572445317), u = j(u, v, w, t, x[o + 6], J, 76029189), t = j(t, u, v, w, x[o + 9], G, 3654602809), w = j(w, t, u, v, x[o + 12], H, 3873151461), v = j(v, w, t, u, x[o + 15], I, 530742520), u = j(u, v, w, t, x[o + 2], J, 3299628645), t = k(t, u, v, w, x[o + 0], K, 4096336452), w = k(w, t, u, v, x[o + 7], L, 1126891415), v = k(v, w, t, u, x[o + 14], M, 2878612391), u = k(u, v, w, t, x[o + 5], N, 4237533241), t = k(t, u, v, w, x[o + 12], K, 1700485571), w = k(w, t, u, v, x[o + 3], L, 2399980690), v = k(v, w, t, u, x[o + 10], M, 4293915773), u = k(u, v, w, t, x[o + 1], N, 2240044497), t = k(t, u, v, w, x[o + 8], K, 1873313359), w = k(w, t, u, v, x[o + 15], L, 4264355552), v = k(v, w, t, u, x[o + 6], M, 2734768916), u = k(u, v, w, t, x[o + 13], N, 1309151649), t = k(t, u, v, w, x[o + 4], K, 4149444226), w = k(w, t, u, v, x[o + 11], L, 3174756917), v = k(v, w, t, u, x[o + 2], M, 718787259), u = k(u, v, w, t, x[o + 9], N, 3951481745), t = c(t, p), u = c(u, q), v = c(v, r), w = c(w, s);
   var O = m(t) + m(u) + m(v) + m(w);
   return O.toLowerCase()
}

js/background.js
var _ctx = 'https://kyfw.12306.cn/otn/'
,	_module = { module : 'login' , rand : 'sjrand' }//图片验证码get请求参数
,	_baTime = '2016-11-15'//选往返程时用到，单程无用
,	_typeArray = {'M' : '一等座' , 'O' : '二等座' , '9' : '商务座' , '1' : '硬座' , '3' : '硬卧' , '4' : '软卧'}

/*辅助变量*/
var _tabId//返回给popup.html信息
,	_isBuyMode//bool模式，buy：立即下单，inquiry：询票当有票时立即下单
,	_initMsgBox = false
,	_is_buy_noSeat//bool是否要买无座票，用于显示
,	_can_noSeat//string,可选无座，当checked是为on否则为undefined
,	_clearTimeout//当用户又再一次提交询票时，清除已有的循环
,	_setTime//number,每隔多少s查询一次票
,	_noShowMsgBox = false//bool，让mosgBox消息框后台运行
,	_errorCount = 0//number,收集msgBox消息框错误提示次数
,	_plugin//object's method imgBase64 => check's Code
,	_buyIfHasTicket = false//后台运行(并自动下单)
,	_user//object 12306用户登录信息
,	_isClickCloseBtnWhenQuery = false//判断是否在询票中，是的话，当点击close按钮时提示后台运行中，有票时询问

/*买票变量*/
var _randCode = ''
,	_secretStr//ticket string
,	_toDate//string出发日期
,	_type//string 'M'/*一等座*/ , 'O'/*二等座*/ , 9/*商务座*/
,	_from//array
,	_to//array
,	_ticketInfo = {}//obj
,	_token
,	_passenger//string
,	_passengerStr//string
,	_time//array，['16:00'班次时间 , '[-1 , 1]'选取上下1小时内的最快班次]
,	_start_time = _old_start_time = ''//string，班次出发时间

window.onload = () => _plugin = document.querySelector('#pluginId')

chrome.runtime.onMessage.addListener(({ match , value } , { id }) => {
    _tabId = id

    if(match === 'userLogin' || match === 'orderInfo') {
		window.clearTimeout(_clearTimeout)
		chrome.notifications.clear('needToBuy')
		chrome.notifications.clear('msgBox')
    }

    if(match === 'userLogin'){
    	_user = value
    	_module = { module : 'login' , rand : 'sjrand' }

		start()
    }else if (match === 'orderInfo') {
		_randCode = _old_start_time = ''
		_module = {
			module : 'passenger',
			rand : 'randp'
		}

		_user = value.user
		_from = value.from
		_to = value.to
		_toDate = value.to_date
		_time = [value.time , /*value.time_range*/]
		_type = value.type

		_passenger = value.name + ',1,'/*证件类型*/ + value.id/*身份证*/ + ','
		//O,0,1,xxx,1,440682199310213272,,N
		_passengerStr = _type + ',0,1,'/*0未知、1成人票*/ + _passenger + value.mobile + ',N'

		_isBuyMode = value.mode === 'buy'
		_can_noSeat = value.can_noSeat
		_setTime = value.loop || 10//默认10s 询票一次
		_noShowMsgBox = _buyIfHasTicket = false
		
		queryTicket()
    }else if (match === 'iconClick') {
    	_noShowMsgBox = false
    }
})

chrome.notifications.onButtonClicked.addListener((id , btnIndex) => {
	chrome.notifications.clear(id)
	if(id === 'needToBuy') ckeckOrderInfo()//由于下单不再需要验证码，所以不用通过start()函数启动
	else if (id === 'msgBox') {
		if(btnIndex === 0) _noShowMsgBox = _buyIfHasTicket = true
		else window.clearTimeout(_clearTimeout)
	}
})

chrome.notifications.onClosed.addListener((id , byUser) => {
	if(id === 'msgBox') {
		_initMsgBox = false

		if(byUser && _isClickCloseBtnWhenQuery) {
			sendMsg(['msgCb' , '后台运行中，有票时询问' , '(点击插件图标重新显示消息框)' , false])

			_noShowMsgBox = true//要放在sendMsg函数之后
			_buyIfHasTicket = false
		}
	}
})

//请求headers 添加Origin、Referer字段
chrome.webRequest.onBeforeSendHeaders.addListener(({ url , requestHeaders }) => {
		Array(
			'loginAysnSuggest',
			'submitOrderRequest',
			localStorage.getItem('leftTicket') || 'queryA',
			'checkOrderInfo',
			'confirmSingleForQueue',
			'checkRandCodeAnsyn',
			'getPassCodeNew',
			'initDc',
			'login/init',
			'getPassengerDTOs',
			'initMy12306',
			'station_name'
		).some(apiName => {
			if(RegExp(apiName).test(url)) {
				return requestHeaders.push({
					name: 'Origin',
					value: 'https://kyfw.12306.cn'
				} , {
					name : 'Referer',
					value : 'https://kyfw.12306.cn/otn/index/init'
				})
			}
		})

		return { requestHeaders }
	},
	{urls: ['*://*/otn/*'] },
	['blocking' , 'requestHeaders']
)

//1、请求验证码
function start (isReLogin/*bool用于重新登录*/){
	let cb = () => dataUrl(val => sendImg2Plugin(val , isReLogin))

	sendMsg(['msgCb' , '等待返回验证码结果...'])

	//删除BIGipServerotn cookie，则图片验证码只需识别一个物品，此时用于登录，
	//提交订单时的图片验证码一定是识别两个物品
	if(_module.module === 'login') {
		chrome.cookies.remove({
			url : 'https://kyfw.12306.cn/otn/passcodeNew/',
			name : 'BIGipServerotn',
		})
		
		Fetch('https://kyfw.12306.cn/otn/login/init' , null , cb , 'text')
	}else cb()
}

//2、验证成功后，登录
function login (cb){
	Fetch(_ctx + 'login/loginAysnSuggest' , {
		'loginUserDTO.user_name' : _user.user,
		'userDTO.password' : _user.pwd,
		'randCode' : _randCode,
	} , () => {
		_module = {
			module : 'passenger',
			rand : 'randp'
		}

		sendMsg(['loginCb' , _user])

		sendMsg(['msgCb' , '登录成功！' , 'login' , false])

		cb && cb()
	})
}

//3、查票
function queryTicket (_undefined){
	//去哪儿询票api
	Fetch('http://train.qunar.com/dict/open/s2s.do?' + ObjStringData({
		dptStation : _from[0],
		arrStation : _to[0],
		date : _toDate,
		type : 'normal',
		user : 'neibu',
		source : 'site',
		start : 1,
		num : 500,
		sort : 3,
	}) , null , res => {
		let returnStr = selectTime(res.data.s2sBeanList)

		if(returnStr === '__query__') {
			Fetch(`${ _ctx }leftTicket/${ localStorage.getItem('leftTicket') || 'queryA' }?${ ObjStringData({//GET
				'leftTicketDTO.train_date':_toDate,
				'leftTicketDTO.from_station':_from[1],
				'leftTicketDTO.to_station':_to[1],
				'purpose_codes':'ADULT',
			}) }` , null , res => {
				res.data.some(({ queryLeftNewDTO : { start_time } , secretStr }) => {
					if(start_time === _start_time) return _secretStr = decodeURIComponent(secretStr)
				})

				_secretStr ? orderTicket() : queryTicket()
			})
		}else if(!_isBuyMode && returnStr === _undefined) {
			//用于判断_start_time是否发生了变化，说明之前班次已停止售票
			if(_old_start_time !== _start_time) {
				if(_old_start_time) {
					_noShowMsgBox = false

					sendMsg(['msgCb' , _old_start_time + '班次已停售，自动调整为' + _start_time , 'queryTicket'])
				}

				_old_start_time = _start_time//放在if判断之后
			}

			_clearTimeout = setTimeout(queryTicket , 1000 * _setTime/*多少秒查一次*/)
		}
	})
}

//4、跳转到提交订单页面
function orderTicket (){
	// chrome.notifications.clear('msgBox')

	Fetch(_ctx + 'leftTicket/submitOrderRequest' , {
		secretStr:_secretStr,
		train_date : _toDate,
		back_train_date:_baTime,
		tour_flag:'dc',
		purpose_codes:'ADULT',
		query_from_station_name:_from[0],
		query_to_station_name:_to[0],
	} , getTicketHtml)
}

//5、获取订单html
function getTicketHtml (){
	Fetch(_ctx + 'confirmPassenger/initDc' , {} , text => {
		let match = text.match(/var ticketInfoForPassengerForm.+\};/)

		if(match) {
			Object.assign(_ticketInfo , JSON.parse(match[0].replace(/(var ticketInfoForPassengerForm=|\;)/g , '').replace(/\'/g , '"')))

			_token = text.match(/var globalRepeatSubmitToken.+\;/)[0].replace(/(.+\s|\'|\;)/g , '')

			let isTimeout = ((new Date(_toDate + ' ' + _start_time).getTime() - new Date(_toDate + ' ' + _time[0]).getTime()) / 1000 / 60) > 30/*大于30分钟时要询问用户是否还买票*/

			_isBuyMode && !isTimeout ? start() : createNotice()
		}else {
			orderTicket()
		}
	} , 'text')
}

//6、先询问一下12306，必须
function ckeckOrderInfo (){
	Fetch(_ctx + 'confirmPassenger/checkOrderInfo' , {
		cancel_flag:2,//未知
		bed_level_order_num:'000000000000000000000000000000',
		passengerTicketStr:_passengerStr,
		oldPassengerStr:_passenger + '3_',
		tour_flag:'dc',
		randCode:_randCode,
		REPEAT_SUBMIT_TOKEN:_token,
	} , buy)
}

//7、下单
function buy (){
	setTimeout(() => {
		Fetch(_ctx + 'confirmPassenger/confirmSingleForQueue' , {
			passengerTicketStr:_passengerStr,
			oldPassengerStr:_passenger + '3_',
			randCode:_randCode,
			purpose_codes:_ticketInfo.purpose_codes,
			key_check_isChange:_ticketInfo.key_check_isChange,
			leftTicketStr:_ticketInfo.leftTicketStr,
			train_location:_ticketInfo.train_location,
			roomType:00,
			dwAll:'N',
			REPEAT_SUBMIT_TOKEN:_token,
		} , ({ data }) => {
			sendMsg(['msgCb' , '下单' + (data.submitStatus ? '成功' : '失败') , JSON.stringify(data)])
			sendMsg(['errorCb'])
		})
	} , 500)
}

//验证图片结果是否正确
checkCode.count = 0//用于统计当msg === ''超过8次后，重新start(isReLogin)
function checkCode (position , isReLogin , isRunSelf){
	_randCode = position.replace(/[()]/g , '')

	!isRunSelf && (checkCode.position = position)

	Fetch(_ctx + 'passcodeNew/checkRandCodeAnsyn' , {
		randCode : _randCode,
		rand : _module.rand
	} , data => {
		let msg = data.data.msg

		//通过记录两个checkCode.isRunSelf值，来判断是否在msg === ''循环时又重新点击checkCode函数
		if(isRunSelf && position !== checkCode.position) return

		if(msg === 'TRUE') {
			sendMsg(['msgCb' , _module.module + 'ing' , 'checkCode'])

			login(isReLogin ? orderTicket : '')
			// setTimeout(() => {
			// 	if(_module.module === 'login') login(isReLogin ? orderTicket : '')
			// 	else ckeckOrderInfo()
			// } , 500)
		}else if(msg === '') {
			++checkCode.count
			checkCode.position = position

			if(checkCode.count > 8) {
				checkCode.count = 0

				start(isReLogin)
			}else {
				checkCode(position , isReLogin , true)
			}
		}else {
			start(isReLogin)
		}
	})
}

//获取验证码图片
function dataUrl (cb){
	var d = new FileReader

	d.onload = () => cb(d.result.replace('data:image/jpeg;base64,' , ''))
	
	Fetch(_ctx + 'passcodeNew/getPassCodeNew?' + ObjStringData(_module) , null , res => d.readAsDataURL(res) , 'blob')
}

//获取验证码结果
function sendImg2Plugin (imgBase64 , isReLogin){
	// Fetch('http://check.huochepiao.360.cn/img_vcode' , JSON.stringify({
	// 	img_buf: imgBase64,
	// 	type: 'D',
	// 	logon: _module.module === 'login' ? 1 : 0,//1代表识别1个物品，0代表2个
	// 	check: _plugin.GetSig(imgBase64)
	// }) , ({ res }) => {
	// 	sendMsg(['msgCb' , 'code：' + res , '获取验证码结果'])
	// 	res ? checkCode(res , isReLogin) : start(isReLogin)
	// })

	/*XMLHttpRequest Origin 有效*/

	var xhr = new XMLHttpRequest
 
	xhr.open('POST' , 'http://check.huochepiao.360.cn/img_vcode' , true)

	xhr.timeout = 3000

	xhr.withCredentials = true//允许携带cookie

	xhr.setRequestHeader('Content-Type' , 'application/x-www-form-urlencoded')

	xhr.onreadystatechange = () => {
	    if (xhr.readyState == 4) {
	    	if(xhr.status == 200) {
	            let res = JSON.parse(xhr.responseText).res
	 
	            sendMsg(['msgCb' , 'code：' + res , '获取验证码结果'])

				res ? setTimeout(() => checkCode(res , isReLogin) , 2500) : start(isReLogin)
	    	}else {
	    		start(isReLogin)
	    	}
	    }
	}

	xhr.send(JSON.stringify({
		img_buf: imgBase64,
		type: 'D',
		logon: _module.module === 'login' ? 1 : 0,//1代表识别1个物品，0代表2个
		check: _plugin.GetSig(imgBase64)
	}))
}

/*---辅助函数---*/
function ObjStringData (data) {
	return Object.keys(data).map((key) => encodeURIComponent(key) + '=' + encodeURIComponent(data[key])).join('&')
}

function Fetch (url , data , cb , resType = 'json'){
	let other = {
		credentials: 'include',
		headers : {
			'Upgrade-Insecure-Requests' : 1//让loginAysnSuggest请求通过
		}
	}
	,	timeout = 4000//timeout 4s

	if(data) {
		other.method = 'POST'
		other.headers['Content-Type'] = 'application/x-www-form-urlencoded;charset=utf-8'

		if(typeof data === 'object') data = ObjStringData(data)

		other.body = data
	}else {
		other.headers['Cache-Control'] = 'no-cache'
	}
	
	sendMsg(['loading'])

	ajax()
	function ajax (){
		//这里使用Promise.race，以最快 resolve 或 reject 的结果来传入后续绑定的回调
		Promise.race([
			fetch(url , other).then(res => res[resType]()),
			new Promise((resolve, reject) => setTimeout(reject , timeout , 'timeout'))
		])
		.then(res => {
			sendMsg(['loading'])

			if(typeof res === 'object') {
				if(res.data && res.data.errMsg) {
					errorHandler(res.data.errMsg)
				}else if(res.messages && res.messages.length !== 0) {
					errorHandler(JSON.stringify(res.messages))
					if(res.messages.indexOf('用户未登录') !== -1) {
						_module = { module : 'login' , rand : 'sjrand' }
						start(true/*需要重新登录*/)
					}else if(res.messages.indexOf('非法请求') !== -1) {
						throw '查票接口更换'
					}
				}
			}

			cb && cb(res)
		} , /*捕获超时错误*/error => {
			errorHandler('请求超时，重新发送中...' , 'timeout')
		})
		.catch(/*捕获请求接口错误*/error => {
			//由于查询余票接口经常改动，所以采用请求github的manifest.json方法获得最新的接口
			if(/leftTicket\/query.+\?/.test(url)) {
				chrome.management.getSelf(({ updateUrl }) => {
					Fetch(updateUrl , null , text => {
						let leftTicket = text.match(new RegExp('</span>update_url<span.+</span>,</td>'))[0].match(/\?(.+)<span/)[1]
						localStorage.setItem('leftTicket' , leftTicket)

						url = url.replace(/query./ , leftTicket)
						ajax()
					} , 'text')
				})
			}else {
				errorHandler('发生未知错误，重新发送中...' , 'timeout')
			}
		})
	}

	function errorHandler (msg , doNext){
		let bool = doNext === 'timeout'

		bool ? ajax() : sendMsg(['errorCb'])

		if(_noShowMsgBox && bool) return

		_noShowMsgBox = false
		sendMsg(['msgCb' , msg , url.slice(25) , false/*无须一直显示*/])
	}
}

//向popup.html发送信息
function sendMsg ([match , value , funcName = '' , isAlwayShow = true]) {
	if(match === 'msgCb') {
		_isClickCloseBtnWhenQuery = !_isBuyMode && funcName === 'selectTime'

		++_errorCount

		!_noShowMsgBox && chrome.notifications[_initMsgBox ? 'update' : 'create']('msgBox' , {
			type : 'basic',
			title : '提示' + _errorCount,
			message : value,
			contextMessage : funcName + '  ' + new Date().getHours() + ':' + new Date().getMinutes(),
			iconUrl : 'logo.png',
			requireInteraction : isAlwayShow,//一直显示
			buttons : _isClickCloseBtnWhenQuery ? [{ title : '后台运行(并自动下单)' } , { title : '停止' }] : [],
		} , () => _initMsgBox = true)
	}else {
		chrome.runtime.sendMessage(_tabId , {
	        match,
	        value,
	    })
	}
}

//询票成功，是否下单
function createNotice (){
	let message
	,	title = _from[0] + ' 到 ' + _to[0] + `  (${ _toDate })   `
	,	buttons = [{ title : '立即购买' }]
	,	otherInfo = _ticketInfo._other

	if(_is_buy_noSeat) {
		message = `(无座票  ${ otherInfo['无座'].price + '元' }  !!!)`
	}else {
		let { price , count } = otherInfo[_typeArray[_type]]

		message = `${ _typeArray[_type] }(  ${ price }元  ) ${ count }张`
	}

	_noShowMsgBox = false

	if(_buyIfHasTicket) {
		_buyIfHasTicket = false
		
		title += '下单中...'
		buttons = []

		//由于下单不再需要验证码，所以不用通过start()函数启动
		ckeckOrderInfo()
	}

	chrome.notifications.create('needToBuy' , {
		type : 'basic',
		title,//广州南 到 北京西(日期)
		message,//硬卧(426.00元)15张票
		iconUrl : 'logo.png',
		contextMessage : otherInfo.trainNo + '   历时' + otherInfo.lishi + '   出发时间' + _start_time,//火车型号G123  历时  出发时间
		requireInteraction : true,//一直显示
		buttons,
	})
}

//选取规定时间内最快班次
function selectTime (data){
	let borderTime = Number(_time[0].replace(/:/ , ''))
	// ,	range = JSON.parse(_time[1])//array
	, 	str
	,	count
	
	data.some(({ saleStatus : { desc } , seats , dptTime , trainNo , extraBeanMap : { interval } } , index) => {
		count = index

		if(desc === '可预订' && borderTime <= Number(dptTime.replace(/:/ , ''))) {
			let bool1 = seats[_typeArray[_type]].count > 0
			,	bool2 = _can_noSeat && seats['无座'].count > 0

			_start_time = dptTime

			if(bool1 || bool2) {
				_is_buy_noSeat = !bool1 && bool2

				str = '__query__'

				seats.trainNo = trainNo
				seats.lishi = interval

				_ticketInfo._other = seats
			}

			return !_isBuyMode || (_isBuyMode && str) ? true : false//return true跳出some循环
		}
	})

	if(str) return str
	else {
		let bool = _isBuyMode || count === data.length - 1

		sendMsg(['msgCb' , bool ? '找不到或今天所有班次无票' : `${ _start_time }班次无票(${ _setTime }秒内重试)` , `selectTime${ bool ? '-nomore' : '' }`])
		
		return bool ? false : str//undefined
	}
}